pystiche/papers/sanakoyeu_et_al_2018/core.py

```diff
@@ -0,0 +1,195 @@
+)
+from .loss import (
+    SanakoyeuEtAl2018GeneratorLoss,
+    DiscriminatorLoss,
+    sanakoyeu_et_al_2018_generator_loss,
+)
+from .data import (
+    sanakoyeu_et_al_2018_dataset,
+    sanakoyeu_et_al_2018_image_loader,
+    sanakoyeu_et_al_2018_images,
+)
+from .utils import sanakoyeu_et_al_2018_optimizer
+
+__all__ = [
+    "sanakoyeu_et_al_2018_training",
+    "sanakoyeu_et_al_2018_stylization",
+    "SanakoyeuEtAl2018Generator",
+    "sanakoyeu_et_al_2018_discriminator",
+    "SanakoyeuEtAl2018TransformerBlock",
+    "sanakoyeu_et_al_2018_generator_loss",
+    "DiscriminatorLoss",
+    "sanakoyeu_et_al_2018_images",
+    "sanakoyeu_et_al_2018_image_loader",
+    "sanakoyeu_et_al_2018_dataset"
+]
+
+
+def default_gan_optim_loop(
+    content_image_loader: DataLoader,
+    style_image_loader: DataLoader,
+    generator: nn.Module,
+    discriminator: nn.Module,
+    transformer_block: nn.Module,
+    discriminator_criterion: nn.Module,
+    generator_criterion: nn.Module,
+    generator_criterion_update_fn: Callable[
+        [nn.Module, nn.Module, torch.tensor, nn.ModuleDict], None
+    ],
+    init_discr_success: float = 0.8,
+    init_win_rate: float = 0.8,
+    alpha: float = 0.05,
+    impl_params: bool = True,
+    get_optimizer: Optional[Callable[[nn.Module], Optimizer]] = None,
+    device: Optional[torch.device] = None,
+) -> nn.Module:
+
+    discriminator_optimizer = get_optimizer(discriminator)
+    generator_optimizer = get_optimizer(generator)
+    style_image_loader = itertools.cycle(style_image_loader)
+
+    discr_success = init_discr_success
+    win_rate = init_win_rate
+
+    def train_discriminator_one_step(fake_image, real_image, discr_success):
+        def closure():
+            discriminator_optimizer.zero_grad()
+            loss = discriminator_criterion(fake_image, real_image)
+            loss.backward()
+            return loss
+
+        discriminator_optimizer.step(closure)
+        acc = discriminator_criterion.get_current_acc()
+        discr_success = discr_success * (1.0 - alpha) + alpha * acc
+
+        return discr_success
+
+    def train_generator_one_step(input_image, discr_success):
+        def closure():
+            generator_optimizer.zero_grad()
+            loss = generator_criterion(input_image)
+            loss.backward()
+            return loss
+
+        generator_optimizer.step(closure)
+        acc = generator_criterion["discr_loss"].get_current_acc()
+        discr_success = discr_success * (1.0 - alpha) + alpha * (1.0 - acc)
+        return discr_success
+
+    for content_image in content_image_loader:
+        content_image = content_image.to(device)
+        stylized_image = generator(content_image)
+
+        if discr_success < win_rate:
+            style_image = next(style_image_loader)
+            style_image = style_image.to(device)
+            if impl_params:
+                fake_images = torch.cat((stylized_image, content_image), 0)
```

Comment:
I think I don't understand the question right now, since I use `sum` for the `loss` and `mean` for the `acc` of the `DiscriminatorOperator`. With `mean` the `acc` is also divided by the `batch_size`, so that the average value of the `acc` is calculated and it is ensured that the `acc` is still between 0 and 1. With `sum`, however, the `loss` of the individual images is summed up, so that in the end I have a sum of the `fake_images` combined here (as would be the case in an individual calculation). 

Or do I still have a mistake in this assumption?

Reply:
Internally you use [`binary_cross_entopy_with_logits`](https://pytorch.org/docs/stable/nn.functional.html#binary-cross-entropy-with-logits) and the default `reduction` is `"mean"`. Thus, if you pass a con`cat`enated image of the size `2xcxHxW` the result is averaged over the two images. If I understand you right you want the sum of the loss of all three images, but right now you are getting

```
total_loss = (loss(fake_image1) + loss(fake_image2)) / 2 + loss(real_image)
```

Reply:
You're completely right, I didn't take that into account in my consideration. Adjusting the `reduction` in the calculation would lead to the correct result, but would get wrong again with another `batch_size`. I just have two different solutions to solve this problem: 
1. Introducing a `loss_correction_factor`, whereby the previous calculation could remain this way, but it would not be clear why this factor exists.
2. An optional third input that would handle these additional images separately. 

If you have comments / thoughts about that feel free to share them.

Reply:
Could you explain or better give an example why `reduction="sum"` would not work in your opinion?

Reply:
The author uses the `reduction="mean"` to calculate the loss that means he uses the average over one `batch`. So a `total_loss` with the following form is needed: 
```python
total_loss = loss(fake_image1)/batch_size + loss(fake_image2)/batch_size + loss(real_image)/batch_size
```
Thus the `loss` with `reduction="mean"` is lower than desired by a factor of 2 and with  `reduction="sum"` higher by a factor depending on the `batch_size`. The exception is of course a `batch_size=1` with `reduction="sum"`. 
Example `batch_size = 2`:
```python
        total_loss = loss(fake_image1)/2 + loss(fake_image2)/2 + loss(real_image)/2
"mean": total_loss = (loss(fake_image1) + loss(fake_image2))/(2+2) + loss(real_image)/2
"sum":  total_loss = loss(fake_image1) + loss(fake_image2) + loss(real_image)/2
```

Reply:
I've completely restructured the `discriminator`. Therefore, this discussion is no longer relevant.